scipy/special/xsf/cdflib.h

```diff
@@ -0,0 +1,91 @@
+
+#pragma once
+
+#include "cephes/igam.h"
+#include "config.h"
+#include "error.h"
+#include "tools.h"
+
+namespace xsf {
+
+XSF_HOST_DEVICE inline double gdtrib(double a, double p, double x) {
+    if (std::isnan(p) || std::isnan(a) || std::isnan(x)) {
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!((0 <= p) && (p <= 1))) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter p is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!(a > 0) || std::isinf(a)) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter a is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (!(x >= 0) || std::isinf(x)) {
+        set_error("gdtrib", SF_ERROR_DOMAIN, "Input parameter x is out of range");
+        return std::numeric_limits<double>::quiet_NaN();
+    }
+    if (x == 0.0) {
+        if (p == 0.0) {
+            set_error("gdtrib", SF_ERROR_DOMAIN, "Indeterminate result for (x, p) == (0, 0).");
+            return std::numeric_limits<double>::quiet_NaN();
+        }
+        /* gdtrib(a, p, x) tends to 0 as x -> 0 when p > 0 */
+        return 0.0;
+    }
+    if (p == 0.0) {
+        /* gdtrib(a, p, x) tends to infinity as p -> 0 from the right when x > 0. */
+        set_error("gdtrib", SF_ERROR_SINGULAR, NULL);
+        return std::numeric_limits<double>::infinity();
+    }
+    if (p == 1.0) {
+        /* gdtrib(a, p, x) tends to 0 as p -> 1.0 from the left when x > 0. */
+        return 0.0;
+    }
+
+    double q = 1.0 - p;
+    double (*func)(double, std::tuple<double, double, double>);
+    std::tuple<double, double, double> args;
+    if (p <= q) {
+        args = {a, p, x};
+        func = +[](double b, std::tuple<double, double, double> args) -> double {
+            auto [a, p, x] = args;
+            return cephes::igam(b, a * x) - p;
+        };
+    } else {
+        args = {a, q, x};
+        func = +[](double b, std::tuple<double, double, double> args) -> double {
+            auto [a, q, x] = args;
+            return q - cephes::igamc(b, a * x);
+        };
+    }
+    double lower_bound = std::numeric_limits<double>::min();
```

Comment:
Why can't you just capture `a, q, x` in the `[]` bracket here?

Reply:
Ah, because you want it as a function pointer due to the if / else? I'd suggest restructuring the code so we don't need `std::tuple`.

Reply:
OK, it turns out performance is the same if you just stick the condition inside `func` like this

```C++
    double q = 1.0 - p;
    auto func = [a, p, q, x](double b) -> double {
	if (p <= q) {
	    return cephes::igam(b, a * x) - p;
	}
	return q - cephes::igamc(b, a * x);
    };
    double lower_bound = std::numeric_limits<double>::min();
    double upper_bound = std::numeric_limits<double>::max();
    auto [xl, xr, f_xl, f_xr, bracket_status] = detail::bracket_root_for_cdf_inversion(
        func, 1.0, lower_bound, upper_bound, -0.875, 7.0, 0.125, 8, false, 342
    );
```

and then there's no need for the function pointer business.

Reply:
Great! You also shouldn't need to specify the return type (the `-> double`) part. That is deduced.

Reply:
It seems like it's needed here.

```
../scipy/special/xsf/cdflib.h: In function ‘double xsf::gdtrib(double, double, double)’:
../scipy/special/xsf/cdflib.h:45:43: error: expected type-specifier before ‘{’ token
   45 |     auto func = [a, p, q, x](double b) -> {
      |                                           ^
ninja: build stopped: subcommand failed.
```

It works if you use `auto` as the type specifier, but in that case, I think I prefer being explicit and just writing `double`.